tape/Ctape_position.man
.\" Copyright (C) 1990-2002 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH CTAPE_POSITION "3castor" "$Date: 2002/04/09 07:44:17 $" CASTOR "Ctape Library Functions"
.SH NAME
Ctape_position \- send a request to the tape daemon to get the tape
positionned to a given file and the HDR labels checked
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "Ctape_api.h"\fR
.sp
.BI "int Ctape_position (char *" path ,
.BI "int " method,
.BI "int " fseq,
.BI "int " fsec,
.BI "unsigned char *" blockid,
.BI "int " Qfirst,
.BI "int " Qlast,
.BI "int " filstat,
.BI "char *" fid,
.BI "char *" fsid,
.BI "char *" recfm,
.BI "int " blksize,
.BI "int " lrecl,
.BI "int " retentd,
.BI "int " flags );
.SH DESCRIPTION
.B Ctape_position
checks if a tape is mounted on the device identified by
.IR path ,
positions it, checks if necessary the HDR1 label and eventually takes the
information from the HDR2 label.
In case of failure the drive is freed.
.TP
.I path
specifies the pathname relative to the current directory or the full pathname
which is used by the application program in the
.B open
statement. This path should have been created by Ctape_mount.
.TP
.I method
may be one of the following:
.RS
.TP
.B TPPOSIT_FSEQ
position by file sequence number
.TP
.B TPPOSIT_FID
position by fid (dataset name)
.TP
.B TPPOSIT_EOI
position at end of information to add a file
.TP
.B TPPOSIT_BLKID
position by block id (locate).
.RE
.TP
.I fseq
file sequence number. This parameter is meaningful only for method
.B TPPOSIT_FSEQ
.TP
.I fsec
file section number. This parameter is meaningful only for methods
.B TPPOSIT_FSEQ
and
.B TPPOSIT_FID
.TP
.I blockid
value returned by a previous Read Block Id
.TP
.I Qfirst
file sequence number corresponding to the first file of the vid mapped volume
.TP
.I Qlast
file sequence number corresponding to the last file of the vid mapped volume
.TP
.I filstat
may be one of the following:
.RS
.TP
.B CHECK_FILE
check file existence and fid in HDR1
.TP
.B NEW_FILE
write a new file without checking file existence or fid
.TP
.B APPEND
append data to an existing tape file
.TP
.B NOCHECKFILE
do not check file existence.
.RE
.TP
.I fid
specifies the fileid in HDR1. If more than 17 characters long,
only the last 17 are taken. It is then converted to upper case.
If
.I filstat
is not
.BR NEW_FILE ,
the fileid parameter is checked against the fid in HDR1.
If
.I filstat
is equal to
.BR NEW_FILE ,
but
.I fid
is NULL,
.I path
is used instead.
.TP
.I fsid
specifies the file set identifier in HDR1. It is the vsn of the first volume of
a volume set.
.TP
.I recfm
.BR F ,
.BR FB ,
.BR FBS ,
.B FS
or
.BR U .
The value is only used to build HDR2.
The default value is taken from the existing tape label if
.I filstat
is
.B APPEND
or
.B CHECK_FILE
or is set to U if
.I filstat
is set to
.BR NEW_FILE .
.TP
.I blksize
specifies the maximum block size. The value is only used to build HDR2.
The default value is taken from the existing tape label if
.I filstat
is
.B APPEND
or
.BR CHECK_FILE .
If
.I filstat
is set to
.BR NEW_FILE ,
for all drive types except Redwoods the default is
32760 if
.I lrecl
is not specified (<= 0),
.I lrecl
otherwise. On Redwoods, the default is 262144 if
.I lrecl
is not specified (<= 0) or the largest multiple of
.I lrecl
below 262144.
.TP
.I lrecl
specifies the logical record length. The value is only used to build HDR2.
The default value is taken from the existing tape label if
.I filstat
is
.B APPEND
or
.BR CHECK_FILE ,
or from the blocksize if
.I filstat
is set to
.BR NEW_FILE .
.TP
.I retentd
retention period in days. A file may be overwritten only if it is expired.
Default is 0, which means that the file may be overwritten immediately.
.TP
.I flags
may be a combination of the following values:
.RS
.TP
.B NOPOS
do not reposition tapes when writing two consecutive files (labelled and
unlabelled tapes) or when reading two consecutive files on unlabelled tapes.
This flag should only be set when the requestor knows that the tape is already
positionned correctly.
.I filstat
must be set to
.BR NEW_FILE .
.TP
.B NOTRLCHK
do not check trailer labels
.TP
.B IGNOREEOI
do not take two consecutive tapemarks as End Of Information.
This option is only valid for label type
.B nl
or
.B blp
and should be avoided on IBM 3420 drives (the reel could wind off).
.RE
.SH RETURN VALUE
This routine returns 0 if the operation was successful or -1 if the operation
failed. In the latter case,
.B serrno
is set appropriately.
.SH ERRORS
.TP 1.2i
.B ENOENT
File not found.
.TP
.B EIO
I/O error.
.TP
.B ENXIO
Drive not operational. The drive is automatically configured down.
.TP
.B EINVAL
Bad parameter value.
.TP
.B SENOSHOST
Host unknown.
.TP
.B SENOSSERV
Service unknown.
.TP
.B SECOMERR
Communication error.
.TP
.B ETDNP
Tape daemon is not running.
.TP
.B ETNRS
Reserve not done.
.TP
.B ETLBL
Bad label structure.
.TP
.B ETFSQ
Bad file sequence number.
.TP
.B ETBLANK
Blank tape.
.TP
.B ETCOMPA
Drive/cartridge compatibility problem.
.TP
.B ETHWERR
Device malfunction.
.TP
.B ETPARIT
Parity error.
.TP
.B ETUNREC
Unrecoverable media error.
.TP
.B ETNOSNS
No sense data available.
.TP
.B ETNRDY
Drive not ready.
.TP
.B ETNXPD
File not expired.
.TP
.B ETNOLBLINFO
Label information not found in memory.
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


tape/Ctape_position.c
/*
 * Copyright (C) 1999-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */

/*	Ctape_position - send a request to the tape daemon to get the tape
 *	positionned to a given file and the HDR labels checked
 */
#include <errno.h>
#include <string.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>
#include "h/Ctape.h"
#include "h/marshall.h"
#include "h/serrno.h"
#include "h/Ctape_api.h"

int Ctape_position(char *path,
                   int method,
                   int fseq,
                   int fsec,
                   unsigned char *blockid,
                   int Qfirst,
                   int Qlast,
                   int filstat,
                   char *fid,
                   char *fsid,
                   char *recfm,
                   int blksize,
                   int lrecl,
                   int retentd,
                   int flags)
{
	char actual_fid[CA_MAXFIDLEN+1];
	int actual_fseq;
	char actual_fsid[CA_MAXVSNLEN+1];
	int c;
	int errflg = 0;
	char func[16];
	char fullpath[CA_MAXPATHLEN+1];
	char *getcwd();
	gid_t gid;
	char hdr1[81], hdr2[81];
	int jid;
	int msglen;
	int n;
	char *p;
	char *q;
	char *rbp;
	char repbuf[328];
	char *sbp;
	char sendbuf[REQBUFSZ];
	char uhl1[81];
	uid_t uid;
	char vol1[81];

	strncpy (func, "Ctape_position", 16);
	uid = getuid();
	gid = getgid();
	jid = findpgrp();

	/* path */

	if (path == NULL || *path == '\0') {
		Ctape_errmsg (func, TP029);
		errflg++;
	} else {
		*fullpath = '\0';
		if (*path != '/') {
			if (getcwd (fullpath, sizeof(fullpath) - 2) == NULL) {
				Ctape_errmsg (func, TP002, "getcwd", strerror(errno));
				errflg++;
			} else {
				strcat (fullpath, "/");
			}
		}
		if (strlen(fullpath) + strlen(path) < sizeof(fullpath)) {
			strcat (fullpath, path);
		} else {
			Ctape_errmsg (func, TP038);
			errflg++;
		}
	}

	if (errflg) {
		serrno = EINVAL;
		return (-1);
	}

        /* Build request header */
 
        sbp = sendbuf;
        marshall_LONG (sbp, TPMAGIC);
        marshall_LONG (sbp, TPPOS);
        q = sbp;        /* save pointer. The next field will be updated */
        msglen = 3 * LONGSIZE;
        marshall_LONG (sbp, msglen);
 
        /* Build request body */
 
	marshall_LONG (sbp, uid);
	marshall_LONG (sbp, gid);
	marshall_LONG (sbp, jid);
	marshall_STRING (sbp, fullpath);
	marshall_WORD (sbp, method);
	if (method == TPPOSIT_FSEQ || method == TPPOSIT_BLKID)
		marshall_LONG (sbp, fseq);
	if (method == TPPOSIT_FSEQ || method == TPPOSIT_FID)
		marshall_WORD (sbp, fsec);
	if (method == TPPOSIT_BLKID)
		marshall_OPAQUE (sbp, blockid, 4);
	marshall_LONG (sbp, Qfirst);
	marshall_LONG (sbp, Qlast);
	marshall_WORD (sbp, filstat);
	if (fid && *fid) {
		p = fid;
		if ((n = strlen (fid) - 17) > 0) p += n;
		strcpy (actual_fid, p);
		UPPER (actual_fid);
		marshall_STRING (sbp, actual_fid);
	} else {
		marshall_STRING (sbp, "");
	}
	if (fsid) {
		strcpy (actual_fsid, fsid);
		UPPER (actual_fsid);
		marshall_STRING (sbp, actual_fsid);
	} else {
		marshall_STRING (sbp, "");
	}
	if (recfm) {
		marshall_STRING (sbp, recfm);
	} else {
		marshall_STRING (sbp, "");
	}
	marshall_LONG (sbp, blksize);
	marshall_LONG (sbp, lrecl);
	marshall_WORD (sbp, retentd);
	marshall_WORD (sbp, flags);

	msglen = sbp - sendbuf;
	marshall_LONG (q, msglen);	/* update length field */

	c = send2tpd (NULL, sendbuf, msglen, repbuf, sizeof(repbuf));
	if (c == 0) {
		rbp = repbuf;
		unmarshall_LONG (rbp, actual_fseq);
		unmarshall_STRING (rbp, vol1);
		unmarshall_STRING (rbp, hdr1);
		unmarshall_STRING (rbp, hdr2);
		unmarshall_STRING (rbp, uhl1);
		c = setlabelinfo (path, flags, actual_fseq, vol1, hdr1, hdr2,
			uhl1);
	}
	return (c);
}
